File: n-ary-tree-preorder-traversal/n-ary-tree-preorder-traversal.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children
"""

class Solution:
    def preorder(self, root: 'Node') -> List[int]:
        def helper(root, res):
            if not root: return []
            
            res.append(root.val)
            if root.children:
                for child in root.children:
                    helper(child, res)
        res = []
        helper(root, res)
        return res

File: n-ary-tree-preorder-traversal/n-ary-tree-preorder-traversal.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val) {
        val = _val;
    }

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution {
public:
    vector<int> preorder(Node* root) {
        if (root == nullptr) return {};
        
        vector<int> res;
        vector<Node*> stack;
        stack.push_back(root);
        
        while (!stack.empty()){
            Node* node = stack[stack.size() - 1];
            stack.pop_back();
            res.push_back(node->val);
            
            for (auto it = node->children.rbegin(); it != node->children.rend(); it++){
                stack.push_back(*it);
            }
        }
        return res;
    }
};